src/PipelineElement.js

```diff
@@ -0,0 +1,139 @@
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @class superclass for pipeline elements
+ *
+ * A pipeline element is the superclass for all classes that can be used
+ * in a pipeline. A pipeline element can be used along with a number of
+ * other pipeline elements of the same type to process a particular type
+ * of file. Each pipeline element should define a type, a name, and
+ * a description.
+ *
+ * @abstract
+ */
+class PipelineElement {
+    /**
+     * Construct a new pipeline element instance.
+     *
+     * @param {Object} [options] options to the constructor
+     * @param {Function} [options.getLogger] a callback function provided by
+     * @param {object} [options.settings] additional settings that can be passed from the
+     * pipeline element to the linter to retrieve the log4js logger
+     */
+    constructor(options) {
+        if (this.constructor === PipelineElement) {
+            throw new Error("Cannot instantiate abstract class PipelineElement directly!");
+        }
+        this.getLogger = options?.getLogger;
+    }
+
+    /**
+     * A callback function provided by the linter to retrieve the log4js logger
+     * @type {Function | undefined}
+     */
+    getLogger;
+
+    /**
+     * Initialize the current plugin.
+     */
+    init() {}
+
+    /**
+     * Name of this type of pipeline element
+     *
+     * Concrete subclasses must define this property.
+     * @readonly
+     * @abstract
+     * @type {string}
+     */
+    // @ts-expect-error: concrete subclasses must define this property
+    name;
+
+    /**
+     * Return the name of this type of pipeline element.
+     * Concrete subclasses must define {@link PipelineElement.name}.
+     *
+     * @returns {String} return the name of this type of pipeline element
+     */
+    getName() {
+        return this.name;
+    }
+
+    /**
+     * Description of what this pipeline element does and what kinds of files it
+     * handles for users who are trying to discover whether or not to use it.
+     *
+     * Concrete subclasses must define this property.
+     * @readonly
+     * @abstract
+     * @type {string}
+     */
+    // @ts-expect-error: concrete subclasses must define this property
+    description;
+
+    /**
+     * Return a description of what this pipeline element does and what kinds of files it
+     * handles for users who are trying to discover whether or not to use it.
+     *
+     * Subclass must define {@link PipelineElement.description}.
+     *
+     * @returns {String} a description of this pipeline element.
+     */
+    getDescription() {
+        return this.description;
+    }
+
+    /**
+     * Type of file that this pipeline element processes. The type should be
+     * a unique name that matches with the type of other pipeline elements that
+     * process this same type of file.
```

Comment:
Could we also indicate that this is used to filter intermediate representations passed into this element? I.e. it must match the type specified in the intermediate representation

Reply:
Updated the description